Explain validation flow in migration guide (#1000)
PrestaShop/docs: src/content/1.7/development/architecture/migration-guide/validation.md

---
title: Validation
weight: 70
---

# Validation

> The purpose of validation is to protect the application from failures, state inconsistency, security related issues etc. Prestashop has different levels of validation:

Reviewer: ```suggestion
The purpose of validation is to protect the application from failures, state inconsistency, security related issues etc. PrestaShop has few different levels of validation:
```

> * **Forms** (the [CRUD]({{<relref "/1.7/development/architecture/migration-guide/forms/crud-forms">}}) or [Settings]({{<relref "/1.7/development/architecture/migration-guide/forms/settings-forms">}})) - are using [Symfony validation constraints](https://symfony.com/doc/3.4/validation.html#constraints) to prevent the user from providing invalid information while also enriching the user experience. For example if a form input expects `Yes` or `No`, but user submitted value `XYZ` we want to reject the data and inform the user what was wrong. The best approach (if possible) is to prevent user providing the wrong value in first place, so in this example we could simply use a checkbox or select input with only 2 possible selections. There are some custom constraints implemented in Prestashop [here](https://github.com/PrestaShop/PrestaShop/tree/1.7.8.x/src/Core/ConstraintValidator).

Reviewer: ```suggestion
* **Forms** (the [CRUD]({{<relref "/1.7/development/architecture/migration-guide/forms/crud-forms">}}) or [Settings]({{<relref "/1.7/development/architecture/migration-guide/forms/settings-forms">}})) - are using [Symfony validation constraints](https://symfony.com/doc/3.4/validation.html#constraints) to prevent the user from providing invalid information while also enriching the user experience. For example if a form input expects `Yes` or `No`, but user submitted value `XYZ` we want to reject the data and inform the user what was wrong. The best approach (if possible) is to prevent user providing the wrong value in the first place, so in this example we could simply use a checkbox or select input with only 2 possible selections. There are some custom constraints implemented in PrestaShop [here](https://github.com/PrestaShop/PrestaShop/tree/1.7.8.x/src/Core/ConstraintValidator).
```

> * [Value objects]({{< relref "/1.7/development/architecture/domain/value_objects" >}}) - after the form is submitted and validated, controller is usually dispatching some `Command`. We must ensure that commands are valid too, because in theory, commands should not be aware whether it was dispatched from a controller or from a command line, therefore it is possible that inputs are not yet validated by the form. The `command` itself should not handle the validation (although there are some older commands that does that, but that might only be some early migration mistakes), but **use the value objects in the constructor instead**. Value objects has their own validation rules and can be reused in any other command. For example, take a look at [ProductCondition](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/Core/Domain/Product/ValueObject/ProductCondition.php) - this is a value object of `Product` domain. It allows constructing specific values and throws `ProductConstraintException` if invalid one is provided. It is currently used in [UpdateProductOptionsCommand](https://github.com/PrestaShop/PrestaShop/1.7.8.x/develop/src/Core/Domain/Product/Command/UpdateProductOptionsCommand.php) and helps to maintain the command validity.

Reviewer: ```suggestion
* [Value objects]({{< relref "/1.7/development/architecture/domain/value_objects" >}}) - after the form is submitted and validated, controller is usually dispatching some `Command`. We must ensure that commands are valid too, because in theory, commands should not be aware whether they were dispatched from a controller or from a command line, therefore it is possible that inputs are not yet validated by the form. The `command` itself should not handle the validation (although there are some older commands that does that, but that might only be some early migration mistakes), but **use the value objects in the constructor instead**. Value objects have their own validation rules and can be reused in any other command. For example, take a look at [ProductCondition](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/Core/Domain/Product/ValueObject/ProductCondition.php) - this is a value object of `Product` domain. It allows constructing specific values and throws `ProductConstraintException` if invalid one is provided. It is currently used in [UpdateProductOptionsCommand](https://github.com/PrestaShop/PrestaShop/1.7.8.x/develop/src/Core/Domain/Product/Command/UpdateProductOptionsCommand.php) and helps to maintain the command validity.
```

* [Domain services]({{< relref "/1.7/development/architecture/domain/domain-services" >}}) - after the command is constructed and dispatched, it reaches the `CommandHandler`, which calls all the needed `Domain services` (there are also lots of `CommandHandlers` that handles all the logic including the validation by themselves, but it is not recommended anymore and should be perceived as migration mistakes). Domain services can perform final checks (e.g. heavy sql queries to check certain records existence) before performing the `write` operations.

> * **Entities** (or the `ObjectModels` in Prestashop) - This is the last level. It is there to prevent data integrity issues (like attempting to store a string into an integer column). This is something that, if all data validations before were done correctly, should rarely happen. It's like **the very last protection against nasty data being persisted**. If these errors still occur, you might get a legacy `PrestaShopException`, but in recent migration process we try to catch it early and replace it by related `CoreException` or `DomainException`. See more in [Domain services]({{< relref "/1.7/development/architecture/domain/domain-services" >}}).

Reviewer 2: ```suggestion
* **Entities** (or the `ObjectModels` in Prestashop) - This is the last level. It is there to prevent data integrity issues (like attempting to store a string into an integer column). This is something that, if all data validations before were done correctly, should rarely happen. This is the last step before data is being persisted: if invalid data has slipped through validation levels and passes this step, then it is persisted in database and might introduce inconsistancy in the application. If these errors still occur, you might get a legacy `PrestaShopException`, but in recent migration process we try to catch it early and replace it by related `CoreException` or `DomainException`. See more in [Domain services]({{< relref "/1.7/development/architecture/domain/domain-services" >}}).
```
I don't know if 'nasty' is good looking in a documentation 😅
